client/lobby/OptionsTab.cpp

```diff
@@ -414,8 +417,355 @@ void OptionsTab::CPlayerOptionTooltipBox::genBonusWindow()
+		drawOutlinedText(x * 57 + 29, y * 63 + 56, (selectedFaction == elem) ? Colors::YELLOW : Colors::WHITE, helper.getName());
+		factions.push_back(elem);
+
+		i++;
+	}
+}
+
+void OptionsTab::SelectionWindow::genContentHeroes()
+{
+	int i = 1;
+
+	// random
+	PlayerSettings set = PlayerSettings();
+	set.hero = -1;
+	CPlayerSettingsHelper helper = CPlayerSettingsHelper(set, SelType::HERO);
+	components.push_back(std::make_shared<CAnimImage>(helper.getImageName(), helper.getImageIndex(), 0, 6, 32 / 2));
+	drawOutlinedText(29, 56, (selectedHero == -1) ? Colors::YELLOW : Colors::WHITE, helper.getName());
+	if(selectedHero == -1)
+		components.push_back(std::make_shared<CPicture>("lobby/townBorderSmallActivated", 6, 32 / 2));
+
+	for(auto & elem : allowedHeroes)
+	{
+		CHero * type = VLC->heroh->objects[elem];
+
+		if(type->heroClass->faction == selectedFaction)
+		{
+
+			int x = i % elementsPerLine;
+			int y = i / elementsPerLine;
+
+			PlayerSettings set = PlayerSettings();
+			set.hero = elem;
+
+			CPlayerSettingsHelper helper = CPlayerSettingsHelper(set, SelType::HERO);
+
+			components.push_back(std::make_shared<CAnimImage>(helper.getImageName(true), helper.getImageIndex(true), 0, x * 57, y * 63));
+			components.push_back(std::make_shared<CPicture>(selectedHero == elem ? "lobby/townBorderBigActivated" : "lobby/townBorderBig", x * 57, y * 63));
+			drawOutlinedText(x * 57 + 29, y * 63 + 56, (selectedHero == elem) ? Colors::YELLOW : Colors::WHITE, helper.getName());
+			heroes.push_back(elem);
+
+			i++;
+		}
+	}
+}
+
+void OptionsTab::SelectionWindow::genContentBonus()
+{
+	PlayerSettings set = PlayerSettings();
+
+	int i = 0;
+	for(auto elem : allowedBonus)
+	{
+		int x = i;
+		int y = 0;
+
+		set.bonus = static_cast<PlayerSettings::Ebonus>(elem);
+		CPlayerSettingsHelper helper = CPlayerSettingsHelper(set, SelType::BONUS);
+		components.push_back(std::make_shared<CAnimImage>(helper.getImageName(), helper.getImageIndex(), 0, x * 57 + 6, y * 63 + 32 / 2));
+		drawOutlinedText(x * 57 + 29, y * 63 + 56, Colors::WHITE , helper.getName());
+		if(selectedBonus == elem)
+		{
+			components.push_back(std::make_shared<CPicture>("lobby/townBorderSmallActivated", x * 57 + 6, y * 63 + 32 / 2));
+			drawOutlinedText(x * 57 + 29, y * 63 + 56, Colors::YELLOW , helper.getName());
+		}
+
+		i++;
+	}
+}
+
+int OptionsTab::SelectionWindow::getElement(const Point & cursorPosition)
+{
+	int x = (cursorPosition.x - pos.x) / 57;
+	int y = (cursorPosition.y - pos.y) / 63;
+
+	return x + y * elementsPerLine;
+}
+
+void OptionsTab::SelectionWindow::clickReleased(const Point & cursorPosition) {
+	int elem = getElement(cursorPosition);
+
+	PlayerSettings set = PlayerSettings();
+	if(type == SelType::TOWN)
+	{
+		if(elem > 0)
+		{
+			elem--;
+			if(elem >= factions.size())
+				return;
+			set.castle = factions[elem];
+		}
+		else
+		{
+			set.castle = -1;
```

Comment:
Does -1 mean random? Is there const for it?

Reply:
set.castle = set.faction?

Reply:
-1 to me means undefined :).
A well named constant would help.

Reply:
> set.castle = set.faction?

Will not change here. Has nothing to do with this PR.

https://github.com/vcmi/vcmi/blob/96e820ddac698c1cc4bb6b31e1908f19c607a945/lib/StartInfo.h#L37C8-L37C8